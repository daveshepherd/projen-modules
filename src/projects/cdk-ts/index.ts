import {
  AwsCdkTypeScriptApp,
  AwsCdkTypeScriptAppOptions,
} from 'projen/lib/awscdk';
import { CdkTypeScriptAppOptions } from './cdk-typescript-app-options';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { Readme } from '../../components/readme';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: CdkTypeScriptAppOptions) {
  const { name } = options;

  const defaults = {
    name,
    defaultReleaseBranch: 'main',
    githubOptions: {
      workflows: false,
    },
    gitignore: ['.npmrc', '.vscode'],
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
    projenrcTs: true,
  } satisfies Partial<CdkTypeScriptAppOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A CDK application in TypeScript
 *
 *
 * @pjid cdk-typescript-app
 */
export class CdkTypeScriptApp extends AwsCdkTypeScriptApp {
  readme: Readme;

  constructor(options: CdkTypeScriptAppOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    } as AwsCdkTypeScriptAppOptions);

    new CodeOwners(this, mergedOptions.codeOwners);
    this.readme = new Readme(this, {
      description: mergedOptions.readme?.description,
    });
    this.readme.addSection(
      'Getting Started',
      '```sh\nyarn install\nnpx projen build\n```',
    );
    if (this.autoMerge) {
      this.github?.mergify?.addRule({
        name: 'Automatic approval for projen upgrade pull requests',
        conditions: [
          'author=endor-projen[bot]',
          ...(this.buildWorkflow?.buildJobIds?.map(
            (id) => `status-success=${id}`,
          ) ?? []),
        ],
        actions: {
          review: {
            type: 'APPROVE',
            message: 'Automatically approving projen upgrade',
          },
        },
      });
      this.github?.mergify?.addRule({
        name: 'Assign PR when check fails',
        conditions: ['#check-failure > 0'],
        actions: {
          assign: {
            add_users: ['daveshepherd'],
          },
        },
      });
    }
  }
}

export * from './cdk-typescript-app-options';
