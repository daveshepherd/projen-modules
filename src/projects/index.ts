export * from './jsii';
export * from './npm';
export * from './python';