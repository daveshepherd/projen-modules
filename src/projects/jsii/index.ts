import { cdk } from 'projen';
import { JsiiProjectOptions } from './jsii-project-options';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: JsiiProjectOptions) {
  const { name } = options;

  const defaults = {
    name,
    autoMerge: false,
    defaultReleaseBranch: 'main',
    gitignore: ['.npmrc', '.vscode'],
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
    projenrcTs: true,
  } satisfies Partial<JsiiProjectOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A JSII project in TypeScript
 *
 *
 * @pjid jsii-project
 */
export class JsiiProject extends cdk.JsiiProject {
  constructor(options: JsiiProjectOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    });

    new CodeOwners(this, mergedOptions.codeOwners);
  }
}

export * from './jsii-project-options';
