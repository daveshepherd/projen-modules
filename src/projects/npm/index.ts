import { typescript } from 'projen';
import { NpmPackageOptions } from './npm-package-options';
import { NpmCircleCi } from '../../circleci';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { Readme } from '../../components/readme';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: NpmPackageOptions) {
  const { name } = options;

  const defaults = {
    name,
    defaultReleaseBranch: 'main',
    githubOptions: {
      workflows: false,
    },
    gitignore: ['.npmrc', '.vscode'],
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
    projenrcTs: true,
  } satisfies Partial<NpmPackageOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A NPM/node package in TypeScript
 *
 *
 * @pjid npm-package
 */
export class NpmPackage extends typescript.TypeScriptProject {
  readme: Readme;

  constructor(options: NpmPackageOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    } as typescript.TypeScriptProjectOptions);

    new CodeOwners(this, mergedOptions.codeOwners);
    new NpmCircleCi(this);
    this.readme = new Readme(this, {
      description: mergedOptions.readme?.description,
    });
    this.readme.addSection(
      'Getting Started',
      '```sh\nyarn install\nnpx projen build\n```',
    );
    if (this.autoMerge) {
      this.github?.mergify?.addRule({
        name: 'Automatic approval for projen upgrade pull requests',
        conditions: [
          'author=endor-projen[bot]',
          ...(this.buildWorkflow?.buildJobIds?.map(
            (id) => `status-success=${id}`,
          ) ?? []),
        ],
        actions: {
          review: {
            type: 'APPROVE',
            message: 'Automatically approving projen upgrade',
          },
        },
      });
    }
  }
}

export * from './npm-package-options';
