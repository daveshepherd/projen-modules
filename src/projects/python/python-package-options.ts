// ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
import type { github, GitOptions, IgnoreFileOptions, javascript, LoggerOptions, Project, ProjectType, ProjenrcJsonOptions, python, RenovatebotOptions, typescript } from 'projen';
import type { ReadmeOptions } from '../../';

/**
 * PythonPackageOptions
 */
export interface PythonPackageOptions {
  /**
   * Venv options.
   * @default - defaults
   * @stability experimental
   */
  readonly venvOptions?: python.VenvOptions;
  /**
   * Use venv to manage a virtual environment for installing dependencies inside.
   * @default - true, unless poetry is true, then false
   * @stability experimental
   * @featured true
   */
  readonly venv?: boolean;
  /**
   * Use setuptools with a setup.py script for packaging and publishing.
   * @default - true, unless poetry is true, then false
   * @stability experimental
   * @featured true
   */
  readonly setuptools?: boolean;
  /**
   * Location of sample tests.
   * Typically the same directory where project tests will be located.
   * @default "tests"
   * @stability experimental
   */
  readonly sampleTestdir?: string;
  /**
   * Include sample code and test if the relevant directories don't exist.
   * @default true
   * @stability experimental
   */
  readonly sample?: boolean;
  /**
   * pytest options.
   * @default - defaults
   * @stability experimental
   */
  readonly pytestOptions?: python.PytestOptions;
  /**
   * Include pytest tests.
   * @default true
   * @stability experimental
   * @featured true
   */
  readonly pytest?: boolean;
  /**
   * Options related to projenrc in TypeScript.
   * @default - default options
   * @stability experimental
   */
  readonly projenrcTsOptions?: typescript.ProjenrcTsOptions;
  /**
   * Use projenrc in TypeScript.
   * This will create a tsconfig file (default: `tsconfig.projen.json`)
   * and use `ts-node` in the default task to parse the project source files.
   * @default false
   * @stability experimental
   */
  readonly projenrcTs?: boolean;
  /**
   * Options related to projenrc in python.
   * @default - default options
   * @stability experimental
   */
  readonly projenrcPythonOptions?: python.ProjenrcOptions;
  /**
   * Use projenrc in Python.
   * This will install `projen` as a Python dependency and add a `synth`
   * task which will run `.projenrc.py`.
   * @default true
   * @stability experimental
   */
  readonly projenrcPython?: boolean;
  /**
   * Options related to projenrc in JavaScript.
   * @default - default options
   * @stability experimental
   */
  readonly projenrcJsOptions?: javascript.ProjenrcOptions;
  /**
   * Use projenrc in javascript.
   * This will install `projen` as a JavaScript dependency and add a `synth`
   * task which will run `.projenrc.js`.
   * @default false
   * @stability experimental
   */
  readonly projenrcJs?: boolean;
  /**
   * Use poetry to manage your project dependencies, virtual environment, and (optional) packaging/publishing.
   * This feature is incompatible with pip, setuptools, or venv.
   * If you set this option to `true`, then pip, setuptools, and venv must be set to `false`.
   * @default false
   * @stability experimental
   * @featured true
   */
  readonly poetry?: boolean;
  /**
   * Use pip with a requirements.txt file to track project dependencies.
   * @default - true, unless poetry is true, then false
   * @stability experimental
   * @featured true
   */
  readonly pip?: boolean;
  /**
   * List of dev dependencies for this project.
   * Dependencies use the format: `<module>@<semver>`
   *
   * Additional dependencies can be added via `project.addDevDependency()`.
   * @default []
   * @stability experimental
   * @featured true
   */
  readonly devDeps?: Array<string>;
  /**
   * List of runtime dependencies for this project.
   * Dependencies use the format: `<module>@<semver>`
   *
   * Additional dependencies can be added via `project.addDependency()`.
   * @default []
   * @stability experimental
   * @featured true
   */
  readonly deps?: Array<string>;
  /**
   * Name of the python package as used in imports and filenames.
   * Must only consist of alphanumeric characters and underscores.
   * @default $PYTHON_MODULE_NAME
   * @stability experimental
   */
  readonly moduleName: string;
  /**
   * Path to the python executable to use.
   * @default "python"
   * @stability experimental
   */
  readonly pythonExec?: string;
  /**
   * Additional fields to pass in the setup() function if using setuptools.
   * @stability experimental
   */
  readonly setupConfig?: Record<string, any>;
  /**
   * Additional options to set for poetry if using poetry.
   * @stability experimental
   */
  readonly poetryOptions?: python.PoetryPyprojectOptionsWithoutDeps;
  /**
   * Package name.
   * @stability experimental
   */
  readonly packageName?: string;
  /**
   * License of this package as an SPDX identifier.
   * @stability experimental
   */
  readonly license?: string;
  /**
   * A URL to the website of the project.
   * @stability experimental
   */
  readonly homepage?: string;
  /**
   * A short description of the package.
   * @stability experimental
   * @featured true
   */
  readonly description?: string;
  /**
   * A list of PyPI trove classifiers that describe the project.
   * @stability experimental
   */
  readonly classifiers?: Array<string>;
  /**
   * Version of the package.
   * @default "0.1.0"
   * @stability experimental
   * @featured true
   */
  readonly version: string;
  /**
   * Author's name.
   * @default $GIT_USER_NAME
   * @stability experimental
   */
  readonly authorName: string;
  /**
   * Author's e-mail.
   * @default $GIT_USER_EMAIL
   * @stability experimental
   */
  readonly authorEmail: string;
  /**
   * Enable VSCode integration.
   * Enabled by default for root projects. Disabled for non-root projects.
   * @default true
   * @stability experimental
   */
  readonly vscode?: boolean;
  /**
   * Auto-close stale issues and pull requests.
   * To disable set `stale` to `false`.
   * @default - see defaults in `StaleOptions`
   * @stability experimental
   */
  readonly staleOptions?: github.StaleOptions;
  /**
   * Auto-close of stale issues and pull request.
   * See `staleOptions` for options.
   * @default false
   * @stability experimental
   */
  readonly stale?: boolean;
  /**
   * Configuration of the README.md file
   */
  readonly readme?: ReadmeOptions;
  /**
   * The name of a secret which includes a GitHub Personal Access Token to be used by projen workflows.
   * This token needs to have the `repo`, `workflows`
   * and `packages` scope.
   * @default "PROJEN_GITHUB_TOKEN"
   * @deprecated use `projenCredentials`
   * @stability deprecated
   */
  readonly projenTokenSecret?: string;
  /**
   * Choose a method of providing GitHub API access for projen workflows.
   * @default - use a personal access token named PROJEN_GITHUB_TOKEN
   * @stability experimental
   */
  readonly projenCredentials?: github.GithubCredentials;
  /**
   * Which type of project this is (library/app).
   * @default ProjectType.UNKNOWN
   * @deprecated no longer supported at the base project level
   * @stability deprecated
   */
  readonly projectType?: ProjectType;
  /**
   * Options for mergify.
   * @default - default options
   * @deprecated use `githubOptions.mergifyOptions` instead
   * @stability deprecated
   */
  readonly mergifyOptions?: github.MergifyOptions;
  /**
   * Whether mergify should be enabled on this repository or not.
   * @default true
   * @deprecated use `githubOptions.mergify` instead
   * @stability deprecated
   */
  readonly mergify?: boolean;
  /**
   * Add a Gitpod development environment.
   * @default false
   * @stability experimental
   */
  readonly gitpod?: boolean;
  /**
   * Options for GitHub integration.
   * @default - see GitHubOptions
   * @stability experimental
   */
  readonly githubOptions?: github.GitHubOptions;
  /**
   * Enable GitHub integration.
   * Enabled by default for root projects. Disabled for non-root projects.
   * @default true
   * @stability experimental
   */
  readonly github?: boolean;
  /**
   * Add a VSCode development environment (used for GitHub Codespaces).
   * @default false
   * @stability experimental
   */
  readonly devContainer?: boolean;
  /**
   * Add a `clobber` task which resets the repo to origin.
   * @default - true, but false for subprojects
   * @stability experimental
   */
  readonly clobber?: boolean;
  /**
   * Configure options for automatic merging on GitHub.
   * Has no effect if
   * `github.mergify` or `autoMerge` is set to false.
   * @default - see defaults in `AutoMergeOptions`
   * @stability experimental
   */
  readonly autoMergeOptions?: github.AutoMergeOptions;
  /**
   * Enable automatic merging on GitHub.
   * Has no effect if `github.mergify`
   * is set to false.
   * @default true
   * @stability experimental
   */
  readonly autoMerge?: boolean;
  /**
   * Enable and configure the 'auto approve' workflow.
   * @default - auto approve is disabled
   * @stability experimental
   */
  readonly autoApproveOptions?: github.AutoApproveOptions;
  /**
   * Options for renovatebot.
   * @default - default options
   * @stability experimental
   */
  readonly renovatebotOptions?: RenovatebotOptions;
  /**
   * Use renovatebot to handle dependency upgrades.
   * @default false
   * @stability experimental
   */
  readonly renovatebot?: boolean;
  /**
   * Options for .projenrc.json.
   * @default - default options
   * @stability experimental
   */
  readonly projenrcJsonOptions?: ProjenrcJsonOptions;
  /**
   * Generate (once) .projenrc.json (in JSON). Set to `false` in order to disable .projenrc.json generation.
   * @default false
   * @stability experimental
   */
  readonly projenrcJson?: boolean;
  /**
   * The shell command to use in order to run the projen CLI.
   * Can be used to customize in special environments.
   * @default "npx projen"
   * @stability experimental
   */
  readonly projenCommand?: string;
  /**
   * The parent project, if this project is part of a bigger project.
   * @stability experimental
   */
  readonly parent?: Project;
  /**
   * The root directory of the project.
   * Relative to this directory, all files are synthesized.
   *
   * If this project has a parent, this directory is relative to the parent
   * directory and it cannot be the same as the parent or any of it's other
   * subprojects.
   * @default "."
   * @stability experimental
   */
  readonly outdir?: string;
  /**
   * Configure logging options such as verbosity.
   * @default {}
   * @stability experimental
   */
  readonly logging?: LoggerOptions;
  /**
   * Configuration options for git.
   * @stability experimental
   */
  readonly gitOptions?: GitOptions;
  /**
   * Configuration options for .gitignore file.
   * @stability experimental
   */
  readonly gitIgnoreOptions?: IgnoreFileOptions;
  /**
   * Whether to commit the managed files by default.
   * @default true
   * @stability experimental
   */
  readonly commitGenerated?: boolean;
  /**
   * This is the name of your project.
   * @default $BASEDIR
   * @stability experimental
   * @featured true
   */
  readonly name: string;
  /**
   * List of teams used to generate the CODEOWNERS file
   */
  readonly codeOwners: Array<string>;
  /**
   * Include a GitHub pull request template.
   * @default true
   */
  readonly pullRequestTemplate?: boolean;
  /**
   * The contents of the pull request template.
   * @default default content
   */
  readonly pullRequestTemplateContents?: Array<string>;
}
