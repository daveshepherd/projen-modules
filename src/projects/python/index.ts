import { PythonProject, PythonProjectOptions } from 'projen/lib/python';
import { PythonPackageOptions } from './python-package-options';
import { CodeOwners } from '../../components/github/codeowners';
import { DEFAULT_PULL_REQUEST_TEMPLATE } from '../../components/github/pull-request-template';
import { Readme } from '../../components/readme';
import { mergeOptions } from '../../utils/merge-options';

function getOptions(options: PythonPackageOptions) {
  const { name } = options;

  const defaults = {
    name,
    pullRequestTemplate: true,
    pullRequestTemplateContents: DEFAULT_PULL_REQUEST_TEMPLATE,
  } satisfies Partial<PythonPackageOptions>;

  return mergeOptions(defaults, options);
}

/**
 * A Python package
 *
 *
 * @pjid python-package
 */
export class PythonPackage extends PythonProject {
  readme: Readme;

  constructor(options: PythonPackageOptions) {
    const mergedOptions = getOptions(options);

    super({
      ...mergedOptions,
    } as PythonProjectOptions);

    new CodeOwners(this, mergedOptions.codeOwners);
    if (mergedOptions.pullRequestTemplate ?? true) {
      this.github?.addPullRequestTemplate(
        ...(mergedOptions.pullRequestTemplateContents ?? []),
      );
    }
    this.readme = new Readme(this, {
      description: mergedOptions.readme?.description,
    });
    this.readme.addSection(
      'Getting Started',
      '```sh\nyarn install\nnpx projen build\n```',
    );
  }
}

export * from './python-package-options';
