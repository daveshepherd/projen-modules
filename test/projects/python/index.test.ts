import { synthSnapshot } from 'projen/lib/util/synth';
import { PythonPackage } from '../../../src/projects/python';

describe('Python Package', () => {
  it('synthesizes', () => {
    const project = new PythonPackage({
      authorEmail: 'test@example.com',
      authorName: 'test',
      codeOwners: ['test'],
      moduleName: 'test',
      name: 'test',
      version: '0.1.0',
    });

    const output = synthSnapshot(project);

    expect(output.CODEOWNERS).toBeDefined();
    expect(output).toMatchSnapshot();
  });
});
