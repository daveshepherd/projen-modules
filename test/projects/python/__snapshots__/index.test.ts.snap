// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Python Package synthesizes 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".

* text=auto eol=lf
/.gitattributes linguist-generated
/.github/pull_request_template.md linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/CODEOWNERS linguist-generated
/requirements-dev.txt linguist-generated
/requirements.txt linguist-generated",
  ".github/pull_request_template.md": "### Background
_Short description of issue being resolved or a feature being added._

**JIRA TICKET**: [SAB-](https://unibuddy.atlassian.net/browse/SAB-)

### Problems Encountered / Decisions Made / Potential Solutions Abandoned
_Any problems you had when working on the PR and what decisions or trade-offs you had to make._

### Check List
* [ ] Added tests for all new code.
* [ ] Added comments and documentation for new code.",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: github.event_name == 'pull_request' || github.event_name == 'pull_request_target'
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
/.env
!/requirements.txt
!/requirements-dev.txt
__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
share/python-wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST
*.manifest
*.spec
pip-log.txt
pip-delete-this-directory.txt
htmlcov/
.tox/
.nox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
*.py,cover
.hypothesis/
.pytest_cache/
cover/
*.mo
*.pot
*.log
local_settings.py
db.sqlite3
db.sqlite3-journal
instance/
.webassets-cache
.scrapy
docs/_build/
.pybuilder/
target/
.ipynb_checkpoints
profile_default/
ipython_config.py
__pypackages__/
celerybeat-schedule
celerybeat.pid
*.sage.py
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/
.spyderproject
.spyproject
.ropeproject
/site
.mypy_cache/
.dmypy.json
dmypy.json
.pyre/
.pytype/
cython_debug/
!/CODEOWNERS
!/.github/pull_request_template.md
!/.projenrc.py
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".",
    "dependencies": [
      {
        "name": "projen",
        "type": "devenv",
        "version": "99.99.99",
      },
      {
        "name": "pytest",
        "type": "test",
        "version": "7.4.3",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".",
    "files": [
      ".gitattributes",
      ".github/pull_request_template.md",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "CODEOWNERS",
      "requirements-dev.txt",
      "requirements.txt",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".",
    "env": {
      "PATH": "$(python -c "import os, sys; python_dir = 'Scripts' if sys.platform == 'win32' else 'bin'; print('{0}/.env/{1}{2}{3}'.format(os.getcwd(), python_dir, os.pathsep, os.environ['PATH']))")",
      "VIRTUAL_ENV": "$(python -c "import os; print('{0}/.env'.format(os.getcwd()))")",
    },
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "clobber": {
        "condition": "git diff --exit-code > /dev/null",
        "description": "hard resets to HEAD of origin and cleans the local repo",
        "env": {
          "BRANCH": "$(git branch --show-current)",
        },
        "name": "clobber",
        "steps": [
          {
            "exec": "git checkout -b scratch",
            "name": "save current HEAD in "scratch" branch",
          },
          {
            "exec": "git checkout $BRANCH",
          },
          {
            "exec": "git fetch origin",
            "name": "fetch latest changes from origin",
          },
          {
            "exec": "git reset --hard origin/$BRANCH",
            "name": "hard reset to origin commit",
          },
          {
            "exec": "git clean -fdx",
            "name": "clean all untracked files",
          },
          {
            "say": "ready to rock! (unpushed commits are under the "scratch" branch)",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
        "steps": [
          {
            "exec": "python .projenrc.py",
          },
        ],
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install and upgrade dependencies",
        "name": "install",
        "steps": [
          {
            "exec": "python -m pip install --upgrade pip",
          },
          {
            "exec": "pip install -r requirements.txt",
          },
          {
            "exec": "pip install -r requirements-dev.txt",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
        "steps": [
          {
            "exec": "pytest",
          },
        ],
      },
    },
  },
  "CODEOWNERS": "* @unibuddy/test",
  "README.md": "# replace this",
  "requirements-dev.txt": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".
projen==99.99.99
pytest==7.4.3
",
  "requirements.txt": "# ~~ Generated by projen. To modify, edit .projenrc.py and run "npx projen".
",
  "test/__init__.py": "__version__ = "0.1.0"
",
  "test/__main__.py": "from .example import hello

if __name__ == "__main__":
    name = input("What is your name? ")
    print(hello(name))
",
  "test/example.py": "def hello(name: str) -> str:
    """A simple greeting.
    Args:
        name (str): Name to greet.
    Returns:
        str: greeting message
    """
    return f"Hello {name}!"
",
  "tests/__init__.py": "",
  "tests/test_example.py": "import pytest

from test.example import hello

@pytest.mark.parametrize(
    ("name", "expected"),
    [
        ("A. Musing", "Hello A. Musing!"),
        ("traveler", "Hello traveler!"),
        ("projen developer", "Hello projen developer!"),
    ],
)
def test_hello(name, expected):
    """Example test with parametrization."""
    assert hello(name) == expected
",
}
`;
